File: search/1_binary_search.zig
const std = @import("std");
const print = std.debug.print;

fn middle(a: []const i32) i32 {
    var mid: i32 = 0;
    if (a.len % 2 == 0) {
        mid = a.len / 2;
    } else {
        mid = (a.len / 2) + 1;
    }
    return mid;
}

pub fn main() !void {
    const a = [_]i32{ 4, 6, 7, 9, 15, 20, 24, 26, 27, 29, 31 };

    var low: usize = 0;
    var high: usize = a.len - 1;

    var mid_index: usize = @divFloor(low + high, 2);
    var mid_value: i32 = a[mid_index];

    const search_num: i32 = 27;

    while (search_num != mid_value) {
        if (search_num > mid_value) {
            low = mid_index + 1;
        } else {
            high = mid_index - 1;
        }

        if (low > high) {
            print("Number not found.\n", .{});
            return;
        }

        mid_index = @divFloor(low + high, 2);
        mid_value = a[mid_index];
    }
    print("Index of {} is: {} \n", .{ search_num, mid_index });
}

File: search/1_binary_search.go
package main

import (
	"fmt"
)

func middle(a []int) int {
	var mid int
	if len(a)%2 == 0 {
		mid = len(a) / 2
	} else {
		mid = (len(a) / 2) + 1
	}
	return mid
}

func binary_search(a []int, searchNum int) {
	low := 0
	high := len(a) - 1

	midIndex := (low + high) / 2
	midValue := a[midIndex]
	found := true

	for searchNum != midValue {
		if searchNum > midValue {
			low = midIndex + 1
		} else {
			high = midIndex - 1
		}

		if low > high {
			found = false
			fmt.Printf("Index of %d is: %d\n", searchNum, -1)
			break
		}

		midIndex = (low + high) / 2
		midValue = a[midIndex]
	}
	if found {

		fmt.Printf("Index of %d is: %d\n", searchNum, midIndex)
	}
}

func main() {
	a := []int{4, 6, 7, 9, 15, 20, 24, 26, 27, 29, 31}
	binary_search(a, 6)
	binary_search(a, 27)
	binary_search(a, 30)

}
